### PR TITLE
User Dashboard

### LINKED ISSUE #11: Dashboard: Friends
As an authenticated user,
I see a section for friends,
In this section, there should be a text field to enter a friend's email and a button to "Add Friend"

**Scenarios**: 
1. If I have not added any friends there should be a message. "You currently have no friends".

2. If I have added friends, I should see a list of all my friends.

**Details:** Users should be able to add a friend by their email address, as long as, the friend is a user of our application and exists in our database. 

**Example:** 
Bugs Bunny and Lola Bunny are users of our application, but Daffy Duck is not. 
1. When Bugs Bunny enters `lola_bunny@gmail.com` to add friend it will be successful and Lola should show up as Bugs Bunny's friend.
2. When Bugs Bunny enters `daffy_duck@gmail.com` to add friend it should give an error message that the user does not exist. 

**Tips:** You'll want to research self-referential `has_many through`. [Here](http://blog.hasmanythrough.com/2007/10/30/self-referential-has-many-through) is a good starting point to understand the concept. You will probably need to do more googling but that's part of the fun ;)

- [x] Write a happy path test 
- [x] Write a sad path test

### LINKED ISSUE #12: Dashboard: Discover Movies
As a authenticated user, 
When I click on the "Discover Movies" button,
I am redirected to a discover page.

### LINKED ISSUE #13: Dashboard
As an authenticated user,
When I visit '/dashboard'
I should see: 
- [x] 'Welcome `<username>`!' at the top of page
- [x] #9  A button to Discover Movies
- [x] #10 A friends section
- [x] #11 A viewing parties section

### PR DESCRIPTION
**Pull request by aschwartz1**
**What Changed**
- Began building user dashboard, closes #13 
  - Button to discover movies closes #12 
  - Friends section which displays friends and allows adding friends, closes #11
  - Viewing parties section, just a skeleton for now (issue #10 exists for the buildout)
- Added/modifed/reorganized routes & controllers for...
  - user dashboard
  - discover page
  - friendship
  - registration
- Added `perform_login` to rails_helper so we can actually login a user in any test
  - NOTE: this could be mocked in tests as an alternative to having capybara actually do a real login
- Added `require_login` helper to ApplicationController, for use as a `before_action` in controllers for pages which require authenticated users

**Next Steps**
- Build out the viewing party section of the dashboard 

**Known Issues**
- Code contains skipped tests for the viewing party section of the dashboard.
  - Added as a roadmap for the viewing party section buildout

### CODE REVIEW

File: app/controllers/friendships_controller.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+      end
+    end
+
+    redirect_to dashboard_path
+  end
+
+  private
+
+  def friend_email
+    params.permit(:friend_email)[:friend_email]
```
Comment (Wil-McC) [inline app/controllers/friendships_controller.rb:22]: curious about this syntax

**Review by Wil-McC — APPROVED** [review]
Great work Alex, the dashboard and dashboard spec are looking super good

File: app/controllers/friendships_controller.rb (hunk 1 above)
Comment (aschwartz1) [reply app/controllers/friendships_controller.rb:22]: Sure!  What about it?

I wrote it like this b/c I only needed the actual email value to use in the controller action. The `params.permit` returns the hash-like params object (IDR the actual object type), and then I'm accessing the `friend_email` key to get the actual email value.

Comment (Wil-McC) [reply app/controllers/friendships_controller.rb:22]: awesome, that answers my question, i wasn't sure what the second :friend_email was doing

**Pull request merged**